src/vs/base/common/arrays.ts

```diff
@@ -588,3 +588,47 @@ export function maxIndex<T>(array: readonly T[], fn: (value: T) => number): numb
 
 	return maxIdx;
 }
+
+/**
+ * Returns an array of all elements `e` with `startKey <= keySelector(e) <= endKey`.
+ * `data` must be sorted in ascending order.
+*/
+export function filterSortedByKey<T>(data: readonly T[], keySelector: (value: T) => number, startKey: number, endKey = startKey): T[] | null {
+	const startIdx = findFirstInSorted(data, x => startKey <= keySelector(x));
+	const endIdx = findFirstInSorted(data, x => endKey < keySelector(x));
+	if (startIdx === endIdx) {
+		return null;
+	}
+	return data.slice(startIdx, endIdx);
+}
+
+/**
+ * This class helps to filter an array by a key when the key decrements for each filter query.
+*/
+export class DecreasingKeyFilter<T> {
+	public static ofSortedDataAsc<T>(data: T[], keySelector: (value: T) => number): DecreasingKeyFilter<T> {
+		return new DecreasingKeyFilter(data, keySelector);
+	}
+
+	private lastIndex = this.data.length - 1;
+
+	constructor(private readonly data: T[], private readonly keySelector: (value: T) => number) {
+	}
+
+	filterByDecreasedKey(startKey: number, endKey = startKey): T[] | null {
```

Comment:
IMHO this is too complicated for its usage outside of tests. Notice how `filterByDecreasedKey` is only called from tests with an `endKey`. I would not complicate things needlessly and eliminate `endKey`. Also, this abstraction is asking a bit too much of the consumer IMHO, this class does not really need to know the key value, as it doesn't sort on the key by itself. It also imposes that a consumer of this class is able to generate numeric keys from its arrays.

I would have modeled this by having a queue and a `takeWhile` method that accepts a predicate as an argument. The contract of `takeWhile` could be that it returns the first elements that match the predicate and it returns empty if the predicate does not hold for the head of the queue.

Reply:
The idea was that you could use `filterByDecreasedKey` like `filter`, with the constraint that the keys are decreasing (not necessarily consecutively) and the elements are sorted by their key to make it fast. It is immediately clear that the usage is correct.

While `takeWhile` is more generic, its usage is also more complicated and does not properly decompose the problem imho. Since the callsite traverses the changed lines in reversed order, the queue would need to be reversed (otherwise you would need a `takeFromBehind`).
Then, you would need to re-reverse the injected texts again so that they are in ascending order.

Also `takeWhile` needs the additional assumption that the traversed lines are compact (i.e. no lines are skipped).

Anyway, I suggest to get rid of `DecreasingKeyFilter` and only use `filterSortedByKey`. If there are multiple injected texts per line, this is actually faster than using `DecreasingKeyFilter`, since `DecreasingKeyFilter` does a linear search to find the end.
After all, I guess the performance mostly depends on the slice.

Reply:
Please continue with what you have chosen and I am happy to review again for correctness.